### DISCUSSION TITLE
fontconfig (system fonts) and lxqt-config-session

### DISCUSSION THREAD

**Discussion by Bluey26** [lxqt/lxqt#2602] [Q&A]
Hello, a few days ago i set up a new archlinux instalation from scatch and i installed lxqt as usual in it, plus the wayland git packages and a few more things.

Today i have found that, for some reason, the `lxqt-config-session` environment variables are not set in here. This seeed to happen to me with the previous instalation too, but i do not know why.

 
![lxqt-config-session](https://github.com/user-attachments/assets/ac517ac2-d6bc-4a20-b503-eb3d06ed5f2f)

This is how my lxqt-config-session looks like, and this is `lxqt-config-locale` (they are both related):

![lxqt-config-locale](https://github.com/user-attachments/assets/fa8e3c8d-7f19-4145-ba75-0f9133782430)

This should write or load every variable mentioned in there into my system, but it does not:
`
echo $LANG `--> `C.UTF-8`

echo `$LC_COLLATE` , echo `$LC_MEASUREMENT` , echo `$LC_NUMERIC` , echo `$LC_TIME `shows empty variables, like if they are not set.

In the other hand, i was having issues in this new setup when i decided to set my default system font. I usually set this using` lxqt-config-appearance ` and i set it to be the `Hack` font.

But for some reason this was not applied, and the only thing that my `~/.config/fontconfig/fonts.conf` got was:

```
<!-- created by lxqt-config-appearance (DO NOT EDIT!) -->
  <include ignore_missing="yes">conf.d</include>
  <match target="font">
    <edit name="antialias" mode="assign">
      <bool>true</bool>
    </edit>
  </match>
  <match target="font">
    <edit name="rgba" mode="assign">
      <const>rgb</const>
    </edit>
  </match>
  <match target="font">
    <edit name="lcdfilter" mode="assign">
      <const>lcddefault</const>
    </edit>
  </match>
  <match target="font">
    <edit name="hinting" mode="assign">
      <bool>true</bool>
    </edit>
  </match>
  <match target="font">
    <edit name="hintstyle" mode="assign">
      <const>hintslight</const>
    </edit>
  </match>
  <match target="font">
    <edit name="autohint" mode="assign">
      <bool>false</bool>
    </edit>
  </match>
  <match target="pattern">
    <edit name="dpi" mode="assign">
      <double>96</double>
    </edit>
  </match>
</fontconfig>

```

You can see in there that no font is set, but some LXQt seemed to recognize "Hack" as my font, and other apps not (i.e mpv was not showing it).

So i came across this whole `fontconfig` thing and after searching a while in the internet i found how to set it up. The problem was that every time i did changes into appearance (lxqt-config-appearance) the `fonts.conf` was overwritten and the font was "unset". So i had to force a `cp` command after closing lxqt-config-appearance to bypass this issue.

Not sure if this was reported before, but since i saw https://github.com/lxqt/lxqt-config/issues/1022 i remember the issue and wanted to comment it.

This font issue, for some reason, was not present in my previous setup, but that one had several years running and i dont know what could have been changed since then to make this happen.

The missing part in the `fonts.conf` file is something like this (in my case):

```
<?xml version='1.0'?>
<!DOCTYPE fontconfig SYSTEM 'fonts.dtd'>
<fontconfig>
 <match target="pattern">
  <test name="family" qual="any">
   <string>sans-serif</string>
  </test>
  <edit mode="assign" name="family" binding="same">
   <string>Hack</string>
  </edit>
 </match>
 <match target="pattern">
  <test name="family" qual="any">
   <string>serif</string>
  </test>
  <edit mode="assign" name="family" binding="same">
   <string>Liberation Serif</string>
  </edit>
 </match>
 <match target="pattern">
  <test name="family" qual="any">
   <string>monospace</string>
  </test>
  <edit mode="assign" name="family" binding="same">
   <string>Liberation Mono</string>
  </edit>
 </match>
```

**PS:** the command `locale` shows:

```
LANG=C.UTF-8
LC_CTYPE="C.UTF-8"
LC_NUMERIC="C.UTF-8"
LC_TIME="C.UTF-8"
LC_COLLATE="C.UTF-8"
LC_MONETARY="C.UTF-8"
LC_MESSAGES="C.UTF-8"
LC_PAPER="C.UTF-8"
LC_NAME="C.UTF-8"
LC_ADDRESS="C.UTF-8"
LC_TELEPHONE="C.UTF-8"
LC_MEASUREMENT="C.UTF-8"
LC_IDENTIFICATION="C.UTF-8"
LC_ALL=
```

**PS2:** The enviroment variables and locale thing seems to be a **wayland** (labwc) specific issue, since in my openbox (X11) session, the variables are loaded and `locale` shows them properly, instead of all the `C.UTF-8` and empty fields.

**Comment by tsujan** [comment]
> I usually set this usinglxqt-config-appearance and i set it to be the Hack font.

The font itself isn't written to `~/.config/fontconfig/fonts.conf` but to `~/.config/lxqt/lxqt.conf`. If you had `Hack` in `fonts.conf` before, you should have manually edited it. Use the subfolder `~/.config/fontconfig/conf.d/` for manual editing instead (see https://github.com/lxqt/qterminal/wiki for an example).

The font is applied correctly here for Qt programs. Sorry, I don't know about GTK apps, but yes, https://github.com/lxqt/lxqt-config/issues/1022 is about them.

**Reply by Bluey26** [reply]
Yes, lxqt.conf shows the font set properly, but for some reason the fonts.conf file was overwritten. (with the first piece of code).

**Comment by tsujan** [comment]
> PS2: The enviroment variables and locale thing seems to be a wayland (labwc) specific issue

So, I think your questions are answered.

As for `lxqt.conf`, please note this line:

```xml
<!-- created by lxqt-config-appearance (DO NOT EDIT!) -->
```

EDIT: We have been aware for several years that LXQt doesn't set the default monospace, emoji,.., but sadly no one has added the feature yet. I think we should have a very old open issue too.

**Reply by stefonarch** [reply]
For testing purposes I use a testuser "user" on tty with `startlxqtwayland` to login and quite often  I remove ~/.config/lxqt` or sometimes  any config files.
 Just tested adding "Hack" and 12px and it was applied immediately everywhere, gtk included, on labwc this.

**Reply by tsujan** [reply]
I think @Bluey26 meant the default(s) for monospace, serif, etc., not the font itself. LXQt doesn't set those defaults; they should be set manually (→ my previous comment).

**Reply by Bluey26** [reply]
> For testing purposes I use a testuser "user" on tty with `startlxqtwayland` to login and quite often I remove ~/.config/lxqt` or sometimes any config files. Just tested adding "Hack" and 12px and it was applied immediately everywhere, gtk included, on labwc this.

Yes, that was what i expected, but for some reason it did not happen in here.

I was fine with having all begin `Hack` font, but i just realized that this function to differentiate the fonts (monospace is used in other type of text).

The problem was that i set the fonts.conf myself but lxqt-config-appearance overwrite it with other text(shown in 1st code) instead of my set fonts (2nd code quote), so i had to add an auxiliary `cp` command to re-overwrite the file every time i open lxqt-config-appearance.

Maybe this will not be an issue with the forementined PR (https://github.com/lxqt/lxqt-config/issues/1022 ) and i could set my fonts.conf accordingly without it being overwritten, but i thought it would be important to mention it.

**Reply by tsujan** [reply]
> The problem was that i set the fonts.conf myself...

As I said, don't do that. LXQt is specific about it, by adding that comment to the beginning of the file.

**Reply by Bluey26** [reply]
Yes, but that does not set the fonts in that file, and thus other programs "don't know" about the lxqt fonts. I try to use Qt based programs, most of them are lxqt-related, but for the other programs it would be as "you dont have a font" because those are not specified in fonts.conf

**Reply by tsujan** [reply]
The font is consistently applied to all Qt apps. If a Qt app doesn't see it, there may be different reasons. For example, it may be a Qt5 app and you may not have libqtxdg-3.12.0 and lxqt-qtplugin-1.4.1, or it may have a hard-coded settings, etc.

"Defaults" is a totally different concept and is defined by your system. It can be changed inside the `~/.config/fontconfig/conf.d` subdirectory.

**Reply by Bluey26** [reply]
Ok, i have to investigate a bit about the "numbers" and the configuration of that directory, to see what is the thing that suits more my needs. Since the lxqt-config-session thing is wayland-only (at least labwc-only) and now is known (or was already known), my questions are answered now, feel free to close this.

**Reply by tsujan** [reply]
> feel free to close this.

Discussions will remain open until eternity ;) They're here to be searched and found by a wider group of users.
